Add documentation for Gardener receiver (#61)
gardener/opentelemetry-collector: receiver/gardenerreceiver/docs/monitoring-a-project.md

# Monitoring a Gardener Project

This document describes how to deploy an OpenTelemetry Collector with the
Gardener receiver to monitor the resources of a single Gardener project. The
examples below describe the deployment of an OpenTelemetry Collector to a Kubernetes
cluster.

## Required Permissions

The receiver only reads from the API server. When only the resources of a
specific project should be monitored, the Gardener
[`viewer` role](https://gardener.cloud/docs/gardener/project/projects/) grants

> the collector's service-account read access to the project's resources.

Reviewer: ```suggestion
the collector's `ServiceAccount` read access to the project's resources.
```

## Prerequisites

* A Kubernetes cluster to deploy the OpenTelemetry Collector to.

> * A service account for a Gardener project with the `viewer` role.

Reviewer: ```suggestion
* A `ServiceAccount` for a Gardener project with the `viewer` role.
```

requires TLS certificates for its admission webhooks; if
[cert-manager](https://cert-manager.io/) is not available in the cluster, let
the chart generate a self-signed certificate instead:

```bash
helm repo add open-telemetry https://open-telemetry.github.io/opentelemetry-helm-charts
helm repo update

helm install opentelemetry-operator open-telemetry/opentelemetry-operator \
  --namespace opentelemetry-operator-system --create-namespace \
  --set admissionWebhooks.certManager.enabled=false \
  --set admissionWebhooks.autoGenerateCert.enabled=true
```

### 2. Create the viewer kubeconfig secret

Create the namespace for the collector and store the project viewer kubeconfig
in a secret. The collector reads the kubeconfig from a file, so the secret is
later mounted into the collector pod.

```bash
kubectl create namespace gardener-monitoring 

kubectl create secret generic gardener-viewer-kubeconfig \
  --namespace gardener-monitoring \
  --from-file=kubeconfig=<PATH_TO_VIEWER_KUBECONFIG>
```

### 3. Deploy the OpenTelemetry Collector

Apply an `OpenTelemetryCollector` resource. It mounts the kubeconfig secret,

> points the receiver at it via the `kubeconfig` option, and restricts the

Reviewer: ```suggestion
points the receiver at it via the kubeconfig option, and restricts the
```

Author: Keeping this since it is a reference to a configuration attribute

spec:
  image: europe-docker.pkg.dev/gardener-project/snapshots/gardener/otel/opentelemetry-collector:latest
  mode: deployment
  replicas: 1
  volumes:
    - name: gardener-viewer-kubeconfig
      secret:
        secretName: gardener-viewer-kubeconfig
  volumeMounts:
    - name: gardener-viewer-kubeconfig
      mountPath: /var/run/secrets/gardener
      readOnly: true
  config:
    receivers:
      gardener:
        kubeconfig: /var/run/secrets/gardener/kubeconfig
        namespace: garden-<PROJECT>
        resources:
          - shoots
    exporters:
      otlp:
        endpoint: <OTLP_BACKEND_ENDPOINT>
    service:
      pipelines:
        metrics:
          receivers: [gardener]
          exporters: [otlp]
```

Save the manifest (for example as `collector.yaml`) and apply it:

```bash
kubectl apply -f collector.yaml
```

The operator reconciles the resource into a collector `Deployment`. Verify that
it is running.

If your OTLP backend requires authentication, please consult the
[OpenTelemetry Collector documentation](https://opentelemetry.io/docs/collector/configuration/#authentication)
for the appropriate configuration.

## Monitor a Gardener Project with Prometheus

> This section deploys the same `gardener` receiver, but exports the metrics via a

Reviewer: ```suggestion
This section deploys the same Gardener receiver, but exports the metrics via a
```

Author: Keeping this since it is a reference to a configuration attribute

          exporters: [prometheus]
```

### 3. Deploy the Prometheus Operator

Install the Prometheus Operator from the `kube-prometheus-stack` Helm chart.
The chart bundles a full monitoring stack, all unnecessary components are disabled
here:

```bash
helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
helm repo update

helm install prometheus-operator prometheus-community/kube-prometheus-stack \
  --namespace monitoring --create-namespace \
  --set prometheus.enabled=false \
  --set alertmanager.enabled=false \
  --set grafana.enabled=false \
  --set defaultRules.enabled=false \
  --set kubeApiServer.enabled=false \
  --set kubeControllerManager.enabled=false \
  --set kubeScheduler.enabled=false \
  --set kubeProxy.enabled=false \
  --set kubeEtcd.enabled=false \
  --set kubeStateMetrics.enabled=false \
  --set nodeExporter.enabled=false \
  --set kubelet.enabled=false \
  --set coreDns.enabled=false \
  --set kubeDns.enabled=false
```

### 4. Create a Prometheus instance and ServiceMonitor

Deploy a `Prometheus` instance with the required RBAC and a `ServiceMonitor`
that scrapes the collector's `prometheus` port. The `Prometheus` instance runs

> under a dedicated service account.

Reviewer: ```suggestion
under a dedicated `ServiceAccount`.
```
